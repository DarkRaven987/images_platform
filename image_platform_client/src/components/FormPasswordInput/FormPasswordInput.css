.form-password-input > label {
  color: #c9c9c9 !important ;
}

.form-password-input {
  display: flex;
  flex-direction: column;
}

.input-with-icon-container {
  position: relative;
  margin-top: 8px;
}

.input-with-icon-container > input {
  width: calc(100% - 20px);
  font-size: 18px;
  padding: 8px;
}

.input-with-icon-container .show-password-icon-container {
  position: absolute;
  right: 16px;
  top: calc(50% - 9px);
  cursor: pointer;
  font-size: 20px;
}

.form-password-input .error-message {
  color: red;
  margin: 10px 0 0;
  font-size: 14px;
  max-width: calc(100% - 20px);
}
