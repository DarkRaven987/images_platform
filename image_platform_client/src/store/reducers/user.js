import { createSlice } from '@reduxjs/toolkit';
import { agent } from '../../utils/agent';
import { API_SIGN_IN_URL, API_SIGN_UP_URL } from '../../utils/agentConsts';

const initialState = null;

const { actions, reducer } = createSlice({
  name: 'user',
  initialState,
  reducers: {
    saveUserData(state, action) {
      return (state = action.payload);
    },
    clearUserData(state, action) {
      return (state = action.payload);
    },
  },
});

export const { saveUserData, clearUserData } = actions;

export const signInAction = async ({ dispatch, username, password }) => {
  if (!username || !password) return false;

  try {
    const userData = await agent.post(API_SIGN_IN_URL, { username, password });
    const { user, accessToken, refreshToken } = userData?.data;

    if (!user?.PK) return false;

    dispatch(saveUserData(user));
    localStorage.setItem('user', JSON.stringify(user));
    localStorage.setItem('accessToken', `Bearer ${accessToken}`);
    localStorage.setItem('refreshToken', `Bearer ${refreshToken}`);

    return true;
  } catch (err) {
    console.error(err);
    return false;
  }
};

export const signUpAction = async ({ username, password }) => {
  if (!username || !password) return false;

  try {
    await agent.post(API_SIGN_UP_URL, { username, password });

    return true;
  } catch (err) {
    console.error(err);
    return false;
  }
};

export const signOutAction = () => {
  localStorage.removeItem('user');
  localStorage.removeItem('accessToken');
  localStorage.removeItem('refreshToken');
  window.location.replace(`${window.location.origin}/login`);
};

export default reducer;
