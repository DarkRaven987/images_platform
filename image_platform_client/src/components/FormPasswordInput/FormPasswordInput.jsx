import { useState, useCallback } from 'react';
import { Visible, NonVisible } from '../../assets/icons';
import './FormPasswordInput.css';

const FormPasswordInput = ({
  value = '',
  onChange = () => {},
  id,
  label,
  className = '',
  style,
  error,
  refer = {},
}) => {
  const { onChange: referOnChange, ...rest } = refer;

  const [val, setVal] = useState(value);
  const [showPassword, setShowPassword] = useState(false);

  const handleChange = useCallback(
    (e) => {
      e.preventDefault();
      setVal(e?.target?.value);
      onChange(e?.target?.value);
      referOnChange && referOnChange(e);
    },
    [onChange, referOnChange],
  );

  const handleClickShowPassword = () => setShowPassword((show) => !show);
  const handleMouseDownPassword = (e) => {
    e.preventDefault();
  };

  return (
    <div
      className={`form-password-input ${className}`}
      style={style}
      data-testid="password-input-container"
    >
      <label htmlFor={id} data-testid="password-input-label">
        {label}
      </label>

      <div className="input-with-icon-container">
        <input
          id={id}
          type={showPassword ? 'text' : 'password'}
          value={val}
          onChange={handleChange}
          data-testid="password-input"
          {...rest}
        />
        <div
          className="show-password-icon-container"
          onClick={handleClickShowPassword}
          onMouseDown={handleMouseDownPassword}
          data-testid="password-input-icon-container"
        >
          {showPassword ? (
            <NonVisible data-testid="password-input-icon-non-visible" />
          ) : (
            <Visible data-testid="password-input-icon-visible" />
          )}
        </div>
      </div>
      {!!error && (
        <p className="error-message" data-testid="password-input-error">
          {error.message}
        </p>
      )}
    </div>
  );
};

export default FormPasswordInput;
