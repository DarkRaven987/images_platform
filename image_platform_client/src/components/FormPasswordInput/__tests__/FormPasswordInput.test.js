import { render, screen, cleanup } from '@testing-library/react';
import user from '@testing-library/user-event';
import FormPasswordInput from '../FormPasswordInput';

const testingText = 'test password';

describe('FormPasswordInput component:', () => {
  beforeEach(() => cleanup());

  it('* renders properly', () => {
    render(<FormPasswordInput value={testingText} />);
    const InputContainer = screen.queryByTestId('password-input-container');
    const InputLabel = screen.queryByTestId('password-input-label');
    const Input = screen.queryByTestId('password-input');
    const InputIconContainer = screen.queryByTestId(
      'password-input-icon-container',
    );
    const InputIconNonVisible = screen.queryByTestId(
      'password-input-icon-non-visible',
    );
    const InputIconVisible = screen.queryByTestId(
      'password-input-icon-visible',
    );
    const InputErrors = screen.queryByTestId('password-input-error');

    expect(InputContainer).toBeTruthy();
    expect(InputLabel).toBeTruthy();
    expect(Input).toBeTruthy();
    expect(InputIconContainer).toBeTruthy();
    expect(InputIconNonVisible).not.toBeTruthy();
    expect(InputIconVisible).toBeTruthy();
    expect(InputErrors).not.toBeTruthy();

    expect(Input.type).toBe('password');

    expect(Input.value).toBe(testingText);
  });

  it('* accepts and applies attributes', () => {
    render(
      <FormPasswordInput
        className="custom-password-class-name"
        id="test-p1"
        label="Test Password"
        value="test_password"
        style={{ color: 'red' }}
        error={{ message: 'Password error test label' }}
      />,
    );
    const InputContainer = screen.queryByTestId('password-input-container');
    const InputLabel = screen.queryByTestId('password-input-label');
    const Input = screen.queryByTestId('password-input');
    const InputErrors = screen.queryByTestId('password-input-error');

    expect(InputContainer.classList).toContain('form-password-input');
    expect(InputContainer.classList).toContain('custom-password-class-name');

    expect(InputLabel.textContent).toBe('Test Password');

    expect(InputErrors).toBeTruthy();

    expect(Input.value).toBe('test_password');
  });

  it('* changes the value', async () => {
    user.setup();

    let test = '';

    render(
      <FormPasswordInput
        className="custom-class-name"
        id="test-1"
        label="Test Input"
        onChange={(val) => (test = val)}
        value={test}
      />,
    );

    const Input = screen.queryByTestId('password-input');

    await user.type(Input, 'User typed here');

    expect(Input.value).toBe(test);
  });

  it('* displays error if provided', () => {
    render(
      <FormPasswordInput
        className="custom-class-name"
        id="test-1"
        label="Test Input"
        type="password"
        error={{
          message: 'Field is required',
        }}
      />,
    );

    const InputErrors = screen.queryByTestId('password-input-error');

    expect(InputErrors).toBeTruthy();

    expect(InputErrors.textContent).toBe('Field is required');
  });

  it('* changes type of field on icon click', async () => {
    user.setup();

    render(<FormPasswordInput value="password" />);

    const Input = screen.queryByTestId('password-input');
    const InputIconContainer = screen.queryByTestId(
      'password-input-icon-container',
    );

    expect(Input.type).toBe('password');

    expect(InputIconContainer).toBeTruthy();
    expect(InputIconContainer).toMatchSnapshot('Icon before icon click');

    await user.click(InputIconContainer);

    expect(Input.type).toBe('text');

    expect(InputIconContainer).toBeTruthy();
    expect(InputIconContainer).toMatchSnapshot('Icon after icon click');

    await user.click(InputIconContainer);

    expect(Input.type).toBe('password');

    expect(InputIconContainer).toBeTruthy();
    expect(InputIconContainer).toMatchSnapshot('Icon before icon click');
  });

  it('* matches the snapshot', async () => {
    user.setup();

    let test = '';

    render(
      <FormPasswordInput
        className="custom-class-name"
        id="test-1"
        label="Test Input"
        onChange={(val) => (test = val)}
        value={test}
      />,
    );
    const InputContainer = screen.queryByTestId('password-input-container');
    const Input = screen.queryByTestId('password-input');

    expect(InputContainer).toMatchSnapshot('Password input before edit');

    await user.type(Input, 'User typed here');

    expect(InputContainer).toMatchSnapshot('Password after before edit');
  });
});
