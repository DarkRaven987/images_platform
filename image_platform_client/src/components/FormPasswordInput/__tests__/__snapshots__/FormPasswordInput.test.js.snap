// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormPasswordInput component: * changes type of field on icon click: Icon after icon click 1`] = `
<div
  class="show-password-icon-container"
  data-testid="password-input-icon-container"
>
  <svg
    data-testid="password-input-icon-non-visible"
  >
    non-visible.svg
  </svg>
</div>
`;

exports[`FormPasswordInput component: * changes type of field on icon click: Icon before icon click 1`] = `
<div
  class="show-password-icon-container"
  data-testid="password-input-icon-container"
>
  <svg
    data-testid="password-input-icon-visible"
  >
    visible.svg
  </svg>
</div>
`;

exports[`FormPasswordInput component: * changes type of field on icon click: Icon before icon click 2`] = `
<div
  class="show-password-icon-container"
  data-testid="password-input-icon-container"
>
  <svg
    data-testid="password-input-icon-visible"
  >
    visible.svg
  </svg>
</div>
`;

exports[`FormPasswordInput component: * matches the snapshot: Password after before edit 1`] = `
<div
  class="form-password-input custom-class-name"
  data-testid="password-input-container"
>
  <label
    data-testid="password-input-label"
    for="test-1"
  >
    Test Input
  </label>
  <div
    class="input-with-icon-container"
  >
    <input
      data-testid="password-input"
      id="test-1"
      type="password"
      value="User typed here"
    />
    <div
      class="show-password-icon-container"
      data-testid="password-input-icon-container"
    >
      <svg
        data-testid="password-input-icon-visible"
      >
        visible.svg
      </svg>
    </div>
  </div>
</div>
`;

exports[`FormPasswordInput component: * matches the snapshot: Password input before edit 1`] = `
<div
  class="form-password-input custom-class-name"
  data-testid="password-input-container"
>
  <label
    data-testid="password-input-label"
    for="test-1"
  >
    Test Input
  </label>
  <div
    class="input-with-icon-container"
  >
    <input
      data-testid="password-input"
      id="test-1"
      type="password"
      value=""
    />
    <div
      class="show-password-icon-container"
      data-testid="password-input-icon-container"
    >
      <svg
        data-testid="password-input-icon-visible"
      >
        visible.svg
      </svg>
    </div>
  </div>
</div>
`;
